Use homekit service callbacks for lights to resolve out of sync states

bdraco opened the pull request:
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Service callbacks allow us to get the on/off, brightness, etc all in one call so we remove all the complexity that was previously needed to handle the out of sync states

We now get the on event and brightness event at the same time which allows us to prevent lights from flashing up to 100% before the requested brightness.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #13695
- This PR is related to issue: 
- Link to documentation pull request: 

Multiple issues #32278 and #13768 and #13695


## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco commented:
Test failure seems unrelated.  Happening in other PRs as well

Jc2k commented:
Nice.

Sorry for the delay getting to this. 

I was worried if some less polished integrations might set their brightness in memory but not actually turn on, but the interface seems to be pretty clear that this should be OK.

My main itch is that it feels weird to put this light specific logic in `HomeDriver`. I can see why, the upstream interface doesn't group the updates so theres no way to implement this on the Accessory as it stands. I wonder if we should propose an amendment upstream to `AccessoryDriver.set_characteristics` to group on `aid` and let each accessory handle the `value` setting part self? (I.e. have an `Accessory.set_characteristics` that gets batches of chars from `AccessoryDriver`). 

Actually that would be a cool refactor because you could coalesce the characteristics writes. E.g. if you received a scene update to hue, saturation, brightness and on you could turn that into a single call to `turn_on`. The current code has to call `turn_on`  for each characteristic write seperately.

bdraco commented:
(I.e. have an `Accessory.set_characteristics` that gets batches of chars from `AccessoryDriver`).
> 
> Actually that would be a cool refactor because you could coalesce the characteristics writes. E.g. if you received a scene update to hue, saturation, brightness and on you could turn that into a single call to `turn_on`. The current code has to call `turn_on` for each characteristic write seperately.

@Jc2k  Thanks for taking a look!

This is a case where AccessoryDriver is providing a bit too much abstraction by only feeding in one change at a time.  My first thought was to change it so you get all the characteristic changes at once from the iOS client and then consume them in batches. The downside is it adds complexity to devices on the Home Assistant side that have to consume the data.

I looked for other special cases where we would want set some characteristic before a device turned on but I didn't find anything besides brightness so I ended up with this solution.  I think it is worth refactoring the interface if there turn out to be more.  You are more familiar with the range of devices than I am so I'll need some guidance on that.

Jc2k commented:
I think we are on the same wavelength re: the abstraction being not quite right.

I think we can change the interface while preserving the current interface, so the other accessories don't have to change. `AccessoryDriver.set_characteristics` would batch the changes and call `Accessory.set_characteristics`. The default impl of `Accessory.set_characteristics` would do what happens now - get the characteristic and update the value. If upstream made that change we could pull it in wihout changing a line of code. But then this PR could target the light accessory class specifically.

I actually think this change could make things *less* complex on the HA side. Consider this bit of code [here](https://github.com/home-assistant/core/blob/dev/homeassistant/components/homekit/type_lights.py#L121-L179). 5 entry points that invoke `turn_on`. `set_saturation` and `set_hue` in particular are weird because HA expects them to be provided as a pair, so it has a weird state machine involving `self._flag` so that it doesnt send a hue without a saturation and vice versa. With the change i'm proposing the self._flag stuff might go away completely.

I think there are other cases lurking around where this refactor would be a win for sure, even if not exactly the same as your original bug:

 * The original bug may well happen for all the other light characteristics - e.g. turn the light on and then change its color, rather than making a single call to `turn_on`.
 * One might be climate. E.g. [here](https://github.com/home-assistant/core/blob/dev/homeassistant/components/homekit/type_thermostats.py#L234-L274). You can see there are 2 largely identical calls to `SERVICE_SET_TEMPERATURE_THERMOSTAT`. The API I suggested could coalesce those service calls together at the very least.

bdraco commented:
After digging in, I think the callbacks should go to the service instead of the characteristic.  I think we could keep the existing api working and register callbacks on the service instead of the characteristic.  That way we could process them all in one transaction.

bdraco commented:
Might be able to do it without upstream changes

1.Remove the callbacks for lights

2. Create a super class for services to allow a setter callback

3. Keep the existing super class. 
```
        # TODO: Add support for chars that do no support notifications.
        for cq in chars_query[HAP_REPR_CHARS]:
            aid, iid = cq[HAP_REPR_AID], cq[HAP_REPR_IID]
            char = self.accessory.get_characteristic(aid, iid)
```
Read all the chars. 
Iterate though self.accessory.services
Iterate though service.characteristics

If char == service.characteristics[...] and service.setter_callback:
 append to our list of chars to send to the service setter callback

then for every service fire service.setter_callback with.  (char_name, value)

```

bdraco commented:
> Might be able to do it without upstream changes
> 
> 1.Remove the callbacks for lights
> 
> 1. Create a super class for services to allow a setter callback
> 2. Keep the existing super class.
> 
> ```
>         # TODO: Add support for chars that do no support notifications.
>         for cq in chars_query[HAP_REPR_CHARS]:
>             aid, iid = cq[HAP_REPR_AID], cq[HAP_REPR_IID]
>             char = self.accessory.get_characteristic(aid, iid)
> ```
> 
> Read all the chars.
> Iterate though self.accessory.services
> Iterate though service.characteristics
> 
> If char == service.characteristics[...] and service.setter_callback:
> append to our list of chars to send to the service setter callback
> 
> then for every service fire service.setter_callback with. (char_name, value)

It would be a lot patching to make that work though so better to see if upstream can accept a patch

bdraco commented:
@Jc2k 

It looks like I can actually do everything in HomeAssistant without patching HAP-python if need be by making a super class for Service. 

https://github.com/bdraco/HAP-python/commit/f0f4ba37847fb3bd5e7facad7194eb0aa601976f

bdraco commented:
@Jc2k Thanks for pointing me at the 5 light entry points.

The per service callback seems to fit a lot better for this use case
```
2020-03-19 02:14:14 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 77}
2020-03-19 02:14:17 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 36}
2020-03-19 02:14:17 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'Brightness': 37}
2020-03-19 02:14:19 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 73}
2020-03-19 02:14:21 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 29}
2020-03-19 02:14:21 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'Brightness': 31}
2020-03-19 02:14:23 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 70}
2020-03-19 02:14:24 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 88}
2020-03-19 02:14:28 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 44}
2020-03-19 02:14:28 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'Brightness': 45}
2020-03-19 02:14:29 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 25}
2020-03-19 02:14:30 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 66}
2020-03-19 02:14:30 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'Brightness': 67}
2020-03-19 02:14:34 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'On': 0}
2020-03-19 02:14:36 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'On': 1}
2020-03-19 02:14:38 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'On': 0}
2020-03-19 02:14:40 DEBUG (Thread-48) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 50}

```

Jc2k commented:
Nice! I was going to suggest doing it on the service originally (HomeKit controller is very service centred). Thought it would be hard to do with the way this integration is written, but adding a service callback is a great idea.

bdraco commented:
@Jc2k. Bought a LIFX bulb to test this with and its working just as well as the native homekit control now.  Thanks again for getting me down this path.  If @ikalchev is ok with https://github.com/ikalchev/HAP-python/pull/229 I'll get this cleaned up and ready for merge

```
2020-03-22 02:07:11 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'Saturation': 67, 'Hue': 30}
2020-03-22 02:07:38 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 68}
2020-03-22 02:07:38 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 72}
2020-03-22 02:07:42 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'On': 0}
2020-03-22 02:07:45 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'Saturation': 23, 'Hue': 28}
2020-03-22 02:08:01 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'On': 0}
2020-03-22 02:08:04 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'Saturation': 67, 'Hue': 30}
2020-03-22 02:08:14 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 68}
2020-03-22 02:08:17 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 78}
2020-03-22 02:08:18 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 97}
2020-03-22 02:08:18 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 100}
2020-03-22 02:08:22 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 56}
2020-03-22 02:08:25 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 34}
2020-03-22 02:08:27 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'On': 1, 'Brightness': 69}
2020-03-22 02:08:29 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'Saturation': 20, 'Hue': 222}
2020-03-22 02:08:30 DEBUG (Thread-8) [custom_components.homekit.type_lights] _set_chars: {'Saturation': 23, 'Hue': 28}
```

Jc2k commented:
Nice! And thank you for getting stuck in with `homekit`. I've barely had time to peer over from `homekit_controller` and help review stuff, so it's nice to see progress again.

bdraco committed:
Switch homekit lights to use service callbacks

Service callbacks allow us to get the on/off, brightness, etc
all in one call so we remove all the complexity that was
previously needed to handle the out of sync states

We now get the on event and brightness event at the same time
which allows us to prevent lights from flashing up to 100%
before the requested brightness.

Jc2k reviewed (approved):
Nice! This looks great!

bdraco commented:
ci failure is unrelated and should be fixed in https://github.com/home-assistant/core/pull/33553

bdraco commented:
rebasing to get a clean ci run to be extra sure this is ok

bdraco commented:
tts test flapped on this run
```
            ] == "{}/api/tts_proxy/42f18378fd4393d18c8dd11d03fa9563c1e54491_de_{}_demo.mp3".format(
```

bdraco commented:
3rd times a charm on running this though CI?

bdraco commented:
Flapped on tts again
```        ] == "{}/api/tts_proxy/42f18378fd4393d18c8dd11d03fa9563c1e54491_de_{}_demo.mp3".format(
```
This has had enough runs for me to be comfortable with it.  Going to recheck coverage before merging though

bdraco commented:
homeassistant/components/homekit/type_lights.py                85      2    98%

bdraco commented:
Missing cover here.  Looking into this

<img width="694" alt="Screen Shot 2020-04-02 at 6 57 25 PM" src="https://user-images.githubusercontent.com/663432/78310517-d6651a80-7513-11ea-93de-3f9f27f403b9.png">

bdraco commented:
> Missing cover here. Looking into this
> 
> <img alt="Screen Shot 2020-04-02 at 6 57 25 PM" width="694" src="https://user-images.githubusercontent.com/663432/78310517-d6651a80-7513-11ea-93de-3f9f27f403b9.png">

Definitely needs coverage for this because it doesn't work as expected if you go from STATE_OFF to STATE_ON with a brightness of 0

bdraco committed:
Fix STATE_OFF -> STATE_ON,brightness:0

bdraco commented:
ok better
```
homeassistant/components/homekit/type_lights.py                84      0   100%
```

bdraco commented:
Ideally this gets fixed in 0.108 since python-HAP was updated to do so.  If someone things its too risky, please remove

FidgetyRat commented:
Really awesome to see the work done on this issue. Thanks to everyone involved, looking forward to seeing it in the main release. 

My wife's eyes will also thank you in the middle of the night.
